{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16161</td>\n",
       "      <td>xqhn0vbp</td>\n",
       "      <td>1e1286db212100993d03cc22374b624f7caee956</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Airborne rhinovirus detection and effect of ul...</td>\n",
       "      <td>10.1186/1471-2458-3-5</td>\n",
       "      <td>PMC140314</td>\n",
       "      <td>12525263.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: Rhinovirus, the most common cause ...</td>\n",
       "      <td>2003-01-13</td>\n",
       "      <td>Myatt, Theodore A; Johnston, Sebastian L; Rudn...</td>\n",
       "      <td>BMC Public Health</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16162</td>\n",
       "      <td>gi6uaa83</td>\n",
       "      <td>8ae137c8da1607b3a8e4c946c07ca8bda67f88ac</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discovering human history from stomach bacteria</td>\n",
       "      <td>10.1186/gb-2003-4-5-213</td>\n",
       "      <td>PMC156578</td>\n",
       "      <td>12734001.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Recent analyses of human pathogens have reveal...</td>\n",
       "      <td>2003-04-28</td>\n",
       "      <td>Disotell, Todd R</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16163</td>\n",
       "      <td>le0ogx1s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A new recruit for the army of the men of death</td>\n",
       "      <td>10.1186/gb-2003-4-7-113</td>\n",
       "      <td>PMC193621</td>\n",
       "      <td>12844350.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>The army of the men of death, in John Bunyan's...</td>\n",
       "      <td>2003-06-27</td>\n",
       "      <td>Petsko, Gregory A</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16164</td>\n",
       "      <td>fy4w7xz8</td>\n",
       "      <td>0104f6ceccf92ae8567a0102f89cbb976969a774</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Association of HLA class I with severe acute r...</td>\n",
       "      <td>10.1186/1471-2350-4-9</td>\n",
       "      <td>PMC212558</td>\n",
       "      <td>12969506.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: The human leukocyte antigen (HLA) ...</td>\n",
       "      <td>2003-09-12</td>\n",
       "      <td>Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16165</td>\n",
       "      <td>0qaoam29</td>\n",
       "      <td>5b68a553a7cbbea13472721cd1ad617d42b40c26</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A double epidemic model for the SARS propagation</td>\n",
       "      <td>10.1186/1471-2334-3-19</td>\n",
       "      <td>PMC222908</td>\n",
       "      <td>12964944.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: An epidemic of a Severe Acute Resp...</td>\n",
       "      <td>2003-09-10</td>\n",
       "      <td>Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine</td>\n",
       "      <td>BMC Infect Dis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47252</td>\n",
       "      <td>16dhu5r9</td>\n",
       "      <td>f5dd44cb2d185bb4af6130c390d584db8c4994bf; f5dd...</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Signatures of Nucleotide Analog Incorporation ...</td>\n",
       "      <td>10.1016/j.celrep.2017.10.005</td>\n",
       "      <td>PMC5670035</td>\n",
       "      <td>29069588.0</td>\n",
       "      <td>cc-by-nc-nd</td>\n",
       "      <td>RNA viruses pose a threat to public health tha...</td>\n",
       "      <td>2017-10-24</td>\n",
       "      <td>Dulin, David; Arnold, Jamie J.; van Laar, Theo...</td>\n",
       "      <td>Cell Rep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47253</td>\n",
       "      <td>ue7g2vpx</td>\n",
       "      <td>79d74f550f25e52d549aa795801c0d8eb8c46dac; 547c...</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Transmission potential of the novel coronaviru...</td>\n",
       "      <td>10.1016/j.idm.2020.02.003</td>\n",
       "      <td>PMC7068636</td>\n",
       "      <td>32190785.0</td>\n",
       "      <td>cc-by-nc-nd</td>\n",
       "      <td>An outbreak of COVID-19 developed aboard the P...</td>\n",
       "      <td>2020-02-29</td>\n",
       "      <td>Mizumoto, Kenji; Chowell, Gerardo</td>\n",
       "      <td>Infect Dis Model</td>\n",
       "      <td>3.006224e+09</td>\n",
       "      <td>#2901</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>noncomm_use_subset</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47272</td>\n",
       "      <td>el1vyyj9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Novel coronavirus takes flight from bats?</td>\n",
       "      <td>10.1038/s41579-020-0336-9</td>\n",
       "      <td>PMC7097671</td>\n",
       "      <td>32051570.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Two recent studies provide initial insights in...</td>\n",
       "      <td>2020-02-12</td>\n",
       "      <td>York, Ashley</td>\n",
       "      <td>Nat Rev Microbiol</td>\n",
       "      <td>3.006140e+09</td>\n",
       "      <td>#713</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47274</td>\n",
       "      <td>7duw51ow</td>\n",
       "      <td>7d77a852039f1cfc2c13843ecfa721c1fe49528c; fc02...</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Lung ultrasound as a diagnostic tool for radio...</td>\n",
       "      <td>10.1016/j.rmed.2017.05.007</td>\n",
       "      <td>PMC5480773</td>\n",
       "      <td>28610670.0</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>BACKGROUND: Pneumonia is a leading cause of mo...</td>\n",
       "      <td>2017-07-03</td>\n",
       "      <td>Ellington, Laura E.; Gilman, Robert H.; Chavez...</td>\n",
       "      <td>Respir Med</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>comm_use_subset</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47291</td>\n",
       "      <td>o96ztrxs</td>\n",
       "      <td>1f2bfec9b985a04dc2645ceabc4e5e456a9ad9a5</td>\n",
       "      <td>PMC</td>\n",
       "      <td>The 2019 Novel Coronavirus: A Crown Jewel of P...</td>\n",
       "      <td>10.2478/jccm-2020-0013</td>\n",
       "      <td>PMC7029402</td>\n",
       "      <td>32104726.0</td>\n",
       "      <td>cc-by-nc-nd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-31</td>\n",
       "      <td>Azamfirei, Razvan</td>\n",
       "      <td>J Crit Care Med (Targu Mures)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#2723</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>noncomm_use_subset</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15861 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid                                                sha source_x  \\\n",
       "16161  xqhn0vbp           1e1286db212100993d03cc22374b624f7caee956      PMC   \n",
       "16162  gi6uaa83           8ae137c8da1607b3a8e4c946c07ca8bda67f88ac      PMC   \n",
       "16163  le0ogx1s                                                NaN      PMC   \n",
       "16164  fy4w7xz8           0104f6ceccf92ae8567a0102f89cbb976969a774      PMC   \n",
       "16165  0qaoam29           5b68a553a7cbbea13472721cd1ad617d42b40c26      PMC   \n",
       "...         ...                                                ...      ...   \n",
       "47252  16dhu5r9  f5dd44cb2d185bb4af6130c390d584db8c4994bf; f5dd...      PMC   \n",
       "47253  ue7g2vpx  79d74f550f25e52d549aa795801c0d8eb8c46dac; 547c...      PMC   \n",
       "47272  el1vyyj9                                                NaN      PMC   \n",
       "47274  7duw51ow  7d77a852039f1cfc2c13843ecfa721c1fe49528c; fc02...      PMC   \n",
       "47291  o96ztrxs           1f2bfec9b985a04dc2645ceabc4e5e456a9ad9a5      PMC   \n",
       "\n",
       "                                                   title  \\\n",
       "16161  Airborne rhinovirus detection and effect of ul...   \n",
       "16162    Discovering human history from stomach bacteria   \n",
       "16163     A new recruit for the army of the men of death   \n",
       "16164  Association of HLA class I with severe acute r...   \n",
       "16165   A double epidemic model for the SARS propagation   \n",
       "...                                                  ...   \n",
       "47252  Signatures of Nucleotide Analog Incorporation ...   \n",
       "47253  Transmission potential of the novel coronaviru...   \n",
       "47272          Novel coronavirus takes flight from bats?   \n",
       "47274  Lung ultrasound as a diagnostic tool for radio...   \n",
       "47291  The 2019 Novel Coronavirus: A Crown Jewel of P...   \n",
       "\n",
       "                                doi       pmcid   pubmed_id      license  \\\n",
       "16161         10.1186/1471-2458-3-5   PMC140314  12525263.0        no-cc   \n",
       "16162       10.1186/gb-2003-4-5-213   PMC156578  12734001.0        no-cc   \n",
       "16163       10.1186/gb-2003-4-7-113   PMC193621  12844350.0        no-cc   \n",
       "16164         10.1186/1471-2350-4-9   PMC212558  12969506.0        no-cc   \n",
       "16165        10.1186/1471-2334-3-19   PMC222908  12964944.0        no-cc   \n",
       "...                             ...         ...         ...          ...   \n",
       "47252  10.1016/j.celrep.2017.10.005  PMC5670035  29069588.0  cc-by-nc-nd   \n",
       "47253     10.1016/j.idm.2020.02.003  PMC7068636  32190785.0  cc-by-nc-nd   \n",
       "47272     10.1038/s41579-020-0336-9  PMC7097671  32051570.0        no-cc   \n",
       "47274    10.1016/j.rmed.2017.05.007  PMC5480773  28610670.0        cc-by   \n",
       "47291        10.2478/jccm-2020-0013  PMC7029402  32104726.0  cc-by-nc-nd   \n",
       "\n",
       "                                                abstract publish_time  \\\n",
       "16161  BACKGROUND: Rhinovirus, the most common cause ...   2003-01-13   \n",
       "16162  Recent analyses of human pathogens have reveal...   2003-04-28   \n",
       "16163  The army of the men of death, in John Bunyan's...   2003-06-27   \n",
       "16164  BACKGROUND: The human leukocyte antigen (HLA) ...   2003-09-12   \n",
       "16165  BACKGROUND: An epidemic of a Severe Acute Resp...   2003-09-10   \n",
       "...                                                  ...          ...   \n",
       "47252  RNA viruses pose a threat to public health tha...   2017-10-24   \n",
       "47253  An outbreak of COVID-19 developed aboard the P...   2020-02-29   \n",
       "47272  Two recent studies provide initial insights in...   2020-02-12   \n",
       "47274  BACKGROUND: Pneumonia is a leading cause of mo...   2017-07-03   \n",
       "47291                                                NaN   2020-01-31   \n",
       "\n",
       "                                                 authors  \\\n",
       "16161  Myatt, Theodore A; Johnston, Sebastian L; Rudn...   \n",
       "16162                                   Disotell, Todd R   \n",
       "16163                                  Petsko, Gregory A   \n",
       "16164  Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...   \n",
       "16165  Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine   \n",
       "...                                                  ...   \n",
       "47252  Dulin, David; Arnold, Jamie J.; van Laar, Theo...   \n",
       "47253                  Mizumoto, Kenji; Chowell, Gerardo   \n",
       "47272                                       York, Ashley   \n",
       "47274  Ellington, Laura E.; Gilman, Robert H.; Chavez...   \n",
       "47291                                  Azamfirei, Razvan   \n",
       "\n",
       "                             journal  Microsoft Academic Paper ID  \\\n",
       "16161              BMC Public Health                          NaN   \n",
       "16162                    Genome Biol                          NaN   \n",
       "16163                    Genome Biol                          NaN   \n",
       "16164                  BMC Med Genet                          NaN   \n",
       "16165                 BMC Infect Dis                          NaN   \n",
       "...                              ...                          ...   \n",
       "47252                       Cell Rep                          NaN   \n",
       "47253               Infect Dis Model                 3.006224e+09   \n",
       "47272              Nat Rev Microbiol                 3.006140e+09   \n",
       "47274                     Respir Med                          NaN   \n",
       "47291  J Crit Care Med (Targu Mures)                          NaN   \n",
       "\n",
       "      WHO #Covidence  has_pdf_parse  has_pmc_xml_parse      full_text_file  \\\n",
       "16161            NaN           True               True      custom_license   \n",
       "16162            NaN           True               True      custom_license   \n",
       "16163            NaN          False               True      custom_license   \n",
       "16164            NaN           True               True      custom_license   \n",
       "16165            NaN           True               True      custom_license   \n",
       "...              ...            ...                ...                 ...   \n",
       "47252            NaN           True               True      custom_license   \n",
       "47253          #2901           True               True  noncomm_use_subset   \n",
       "47272           #713          False               True      custom_license   \n",
       "47274            NaN           True               True     comm_use_subset   \n",
       "47291          #2723           True               True  noncomm_use_subset   \n",
       "\n",
       "                                                     url  \n",
       "16161  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "16162  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "16163  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "16164  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "16165  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "...                                                  ...  \n",
       "47252  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...  \n",
       "47253  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...  \n",
       "47272  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...  \n",
       "47274  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...  \n",
       "47291  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...  \n",
       "\n",
       "[15861 rows x 18 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df.sha.fillna(\"\", inplace=True)\n",
    "df = pd.read_csv(\"./CORD-19-research-challenge/metadata.csv\")\n",
    "df[df.has_pmc_xml_parse == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "root_folder = \"CORD-19-research-challenge\"\n",
    "#get full text\n",
    "def get_full_text(paper):\n",
    "    license = paper[\"full_text_file\"]\n",
    "    if paper[\"has_pdf_parse\"]:\n",
    "        paper_type = \"pdf_json\"\n",
    "        file_names = str(paper.sha).split(\"; \")\n",
    "        suffix = \".json\"\n",
    "    elif paper[\"has_pmc_xml_parse\"]:\n",
    "        paper_type = \"pmc_json\"\n",
    "        file_names = str(paper.pmcid).split(\"; \")\n",
    "        suffix = \".xml.json\"\n",
    "    else:\n",
    "        return \"\"\n",
    "    full_text = []\n",
    "    for file in file_names:\n",
    "        file_path = os.path.join(root_folder, license, license, paper_type, file+suffix)\n",
    "        with open(file_path) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            for section in [\"abstract\", \"body_text\"]:\n",
    "                if section in data:\n",
    "                    for paragraph in data[section]:\n",
    "                        full_text.append(paragraph[\"text\"])\n",
    "            \n",
    "    return \" \".join(full_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>8q5ondtn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Intrauterine virus infections and congenital h...</td>\n",
       "      <td>10.1016/0002-8703(72)90077-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4361535.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract The etiologic basis for the vast majo...</td>\n",
       "      <td>1972-12-31</td>\n",
       "      <td>Overall, James C.</td>\n",
       "      <td>American Heart Journal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-8703(72)90077-4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>pzfd0e50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Coronaviruses in Balkan nephritis</td>\n",
       "      <td>10.1016/0002-8703(80)90355-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6243850.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1980-03-31</td>\n",
       "      <td>Georgescu, Leonida; Diosi, Peter; Buţiu, Ioan;...</td>\n",
       "      <td>American Heart Journal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-8703(80)90355-5</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>22bka3gi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Cigarette smoking and coronary heart disease: ...</td>\n",
       "      <td>10.1016/0002-8703(80)90356-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7355701.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1980-03-31</td>\n",
       "      <td>Friedman, Gary D</td>\n",
       "      <td>American Heart Journal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-8703(80)90356-7</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>zp9k1k3z</td>\n",
       "      <td>aecbc613ebdab36753235197ffb4f35734b5ca63</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Clinical and immunologic studies in identical ...</td>\n",
       "      <td>10.1016/0002-9343(73)90176-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4579077.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Middle-aged female identical twins, o...</td>\n",
       "      <td>1973-08-31</td>\n",
       "      <td>Brunner, Carolyn M.; Horwitz, David A.; Shann,...</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(73)90176-9</td>\n",
       "      <td>Middle-aged female identical twins, one of who...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>cjuzul89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Epidemiology of community-acquired respiratory...</td>\n",
       "      <td>10.1016/0002-9343(85)90361-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4014285.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Upper respiratory tract infections ar...</td>\n",
       "      <td>1985-06-28</td>\n",
       "      <td>Garibaldi, Richard A.</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(85)90361-4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha  source_x  \\\n",
       "0  8q5ondtn                                       NaN  Elsevier   \n",
       "1  pzfd0e50                                       NaN  Elsevier   \n",
       "2  22bka3gi                                       NaN  Elsevier   \n",
       "3  zp9k1k3z  aecbc613ebdab36753235197ffb4f35734b5ca63  Elsevier   \n",
       "4  cjuzul89                                       NaN  Elsevier   \n",
       "\n",
       "                                               title  \\\n",
       "0  Intrauterine virus infections and congenital h...   \n",
       "1                  Coronaviruses in Balkan nephritis   \n",
       "2  Cigarette smoking and coronary heart disease: ...   \n",
       "3  Clinical and immunologic studies in identical ...   \n",
       "4  Epidemiology of community-acquired respiratory...   \n",
       "\n",
       "                            doi pmcid  pubmed_id    license  \\\n",
       "0  10.1016/0002-8703(72)90077-4   NaN  4361535.0  els-covid   \n",
       "1  10.1016/0002-8703(80)90355-5   NaN  6243850.0  els-covid   \n",
       "2  10.1016/0002-8703(80)90356-7   NaN  7355701.0  els-covid   \n",
       "3  10.1016/0002-9343(73)90176-9   NaN  4579077.0  els-covid   \n",
       "4  10.1016/0002-9343(85)90361-4   NaN  4014285.0  els-covid   \n",
       "\n",
       "                                            abstract publish_time  \\\n",
       "0  Abstract The etiologic basis for the vast majo...   1972-12-31   \n",
       "1                                                NaN   1980-03-31   \n",
       "2                                                NaN   1980-03-31   \n",
       "3  Abstract Middle-aged female identical twins, o...   1973-08-31   \n",
       "4  Abstract Upper respiratory tract infections ar...   1985-06-28   \n",
       "\n",
       "                                             authors  \\\n",
       "0                                  Overall, James C.   \n",
       "1  Georgescu, Leonida; Diosi, Peter; Buţiu, Ioan;...   \n",
       "2                                   Friedman, Gary D   \n",
       "3  Brunner, Carolyn M.; Horwitz, David A.; Shann,...   \n",
       "4                              Garibaldi, Richard A.   \n",
       "\n",
       "                            journal  Microsoft Academic Paper ID  \\\n",
       "0            American Heart Journal                          NaN   \n",
       "1            American Heart Journal                          NaN   \n",
       "2            American Heart Journal                          NaN   \n",
       "3  The American Journal of Medicine                          NaN   \n",
       "4  The American Journal of Medicine                          NaN   \n",
       "\n",
       "  WHO #Covidence  has_pdf_parse  has_pmc_xml_parse  full_text_file  \\\n",
       "0            NaN          False              False  custom_license   \n",
       "1            NaN          False              False  custom_license   \n",
       "2            NaN          False              False  custom_license   \n",
       "3            NaN           True              False  custom_license   \n",
       "4            NaN          False              False  custom_license   \n",
       "\n",
       "                                            url  \\\n",
       "0  https://doi.org/10.1016/0002-8703(72)90077-4   \n",
       "1  https://doi.org/10.1016/0002-8703(80)90355-5   \n",
       "2  https://doi.org/10.1016/0002-8703(80)90356-7   \n",
       "3  https://doi.org/10.1016/0002-9343(73)90176-9   \n",
       "4  https://doi.org/10.1016/0002-9343(85)90361-4   \n",
       "\n",
       "                                                text  \n",
       "0                                                     \n",
       "1                                                     \n",
       "2                                                     \n",
       "3  Middle-aged female identical twins, one of who...  \n",
       "4                                                     "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['text'] = df.apply(lambda x: get_full_text(x), axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'shows the phylogeny reconstructed using the neighboring-joining method with mid-point rooting. One striking feature is that GZ01 was distantly separated from other sequences. All the sequences from mainland China (GZ01 and BJ01-04) clustered together and separated from the remaining sequences. Interestingly, CUHK-W1 shared some common variants with those from mainland China, suggesting multiple sources of SARS-CoV in Hong Kong. A maximum parsimony tree was similar to that in Figure S1 . The phylogeny in Figure replicates. 2 Five Singaporean isolates (SIN2500, SIN2677, SIN2679, SIN2748, and SIN2774): SIN2500 was the primary index patient in Singapore and directly linked to index patient A in Hong Kong. SIN2677, SIN 2748, SIN2774, and SINxxxx (this patient was unavailable) were the four primary contacts and SIN2679 was the secondary contact of SINxxxx [9] . For those isolates, all patients have \"history of close contact with another patient with SARS or travel to a region with documented community transmission of SARS within 10 days of onset of symptoms\", and clinical samples are \"obtained form the patients between 0 and 11 days after of onset of symptoms\" (see [9] ). Therefore, the time for the SIN2500 was estimated to be between 7 days (mean incubation time) and 21 (10 + 11) days. Because of the close clinical relationship, we estimated the minimum incubation period 2 days [10] for SIN2677, SIN2748, SIN2774, and SINxxxx, and maximum 13 (2 + 11) days. For SIN2679, the estimated range is likewise to be 2 and 13 days from SINxxxx. Before we finished the data analysis, we did not find a formal report of the clinical relationship and sample isolation dates for these three isolates. The specimens were probably isolated before April 2, two weeks before the sequences were released in mid-April, 2003. Therefore, we used the divergence time from 12 days (7 incubation days + 5 hospital admission days) to 41 days The estimated divergence time between each pair of sequences is shown in Table S1 and S2 and the proportion of nucleotide difference is shown in Table S3 .  The outbreak of severe acute respiratory syndrome (SARS) caused a severe global epidemic in 2003 which led to hundreds of deaths and many thousands of hospitalizations. The virus causing SARS was identified as a novel coronavirus (SARS-CoV) and multiple genomic sequences have been revealed since mid-April, 2003. After a quiet summer and fall in 2003, the newly emerged SARS cases in Asia, particularly the latest cases in China, are reinforcing a wide-spread belief that the SARS epidemic would strike back. With the understanding that SARS-CoV might be with humans for years to come, knowledge of the evolutionary mechanism of the SARS-CoV, including its mutation rate and emergence time, is fundamental to battle this deadly pathogen. To date, the speed at which the deadly virus evolved in nature and the elapsed time before it was transmitted to humans remains poorly understood. Results: Sixteen complete genomic sequences with available clinical histories during the SARS outbreak were analyzed. After careful examination of multiple-sequence alignment, 114 single nucleotide variations were identified. To minimize the effects of sequencing errors and additional mutations during the cell culture, three strategies were applied to estimate the mutation rate by 1) using the closely related sequences as background controls; 2) adjusting the divergence time for cell culture; or 3) using the common variants only. The mutation rate in the SARS-CoV genome was estimated to be 0.80 -2.38 × 10 -3 nucleotide substitution per site per year which is in the same order of magnitude as other RNA viruses. The non-synonymous and synonymous substitution rates were estimated to be 1.16 -3.30 × 10 -3 and 1.67 -4.67 × 10 -3 per site per year, respectively. The most recent common ancestor of the 16 sequences was inferred to be present as early as the spring of 2002. The estimated mutation rates in the SARS-CoV using multiple strategies were not unusual among coronaviruses and moderate compared to those in other RNA viruses. All estimates of mutation rates led to the inference that the SARS-CoV could have been with humans in the spring of 2002 without causing a severe epidemic. The earliest confirmed case of the severe acute respiratory syndrome (SARS) occurred in November, 2002 in the Guangdong province of China. Toward the end of the epidemic (as reported by July 31, 2003) there were 8,098 recognized cases in 31 countries or regions worldwide and 774 implicated deaths (WHO, http://www.who.int/csr/ sars/country/table2003_09_23/en/). Due to an unprecedented international effort, the SARS coronavirus (SARS-CoV) was identified as the causal agent in late March 2003 and its first complete genomic sequences were published April 13, 2003 [1, 2] . One month later, SARS-like coronaviruses were found in palm civets and other animals in Guangdong, China, the first evidence of possible interspecies transmission of the virus [3] . The re-emergence of the isolated SARS cases in Asia in December, 2003 and in Anhui province and Beijing, China, in late April 2004, has confirmed a wide-spread conjecture that the SARS-CoV will likely be with humans for years to come. This reemergence of SARS cases makes it legitimate to critically re-evaluate the time for the origin of the SARS-CoV. There are 26 putative coding regions which cover about 98% of the 29.8-kb SARS-CoV genome. Approximately two-thirds of the genome are at the 5\\' side encoding the nonstructural proteins (orf1ab and orf1a) and one-third are at the 3\\' side encoding four structural proteins: spike glycoprotein (S), envelope (E), membrane (M), and nucleocapsid (N) [4] . The spike glycoprotein, especially its S1 subdomain, is responsible for binding to the specific receptor in the target cells [4, 5] . RNA polymerase and nsp1 genes are two major loci in orf1ab. Estimating the mutation rate in RNA viruses and retroviruses is critical but also challenging for tracing their rapidly evolving paths. The rates estimated from the positivestrand ssRNA virus appear to be in a similar range (e.g., 10 -3 per site per year) from the negative-strand ssRNA virus, although a direct comparison is not possible because the mutation rates could be estimated from different regions or genes [6] [7] [8] [9] [10] [11] [12] [13] [14] [15] . The estimated mutation rates in coronavirus, which SARS-CoV phylogenetically links to, are moderate to high compared to the others in the category of ssRNA viruses. For example, it was estimated to be 0.3 -0.6 × 10 -2 per site per year in the infectious bronchitis virus in a previous study [8] . However, the estimated mutation rate appears to have a wider range in the retrovirus [16] [17] [18] [19] [20] . More details are presented in the Discussion section. How SARS-CoV evolves has important implications for both strategic planning in the prevention of SARS epidemics and development of a vaccine and antibodies. The mutation rate is among the most fundamental aspect of sequence evolution. If the pathogen evolves slowly, there will be a better chance for development of effective long lasting vaccines and successful treatment for patients from a particular geographic region will likely be effective for patients from other areas. On the other hand, if the pathogen (particularly the genes coding for major antigens) evolves rapidly, an effective strategy to prevent transmission of the SARS-CoV must be the top-priority, and an effective vaccine program may be problematic. The purpose of this study is to improve our understanding of the evolutionary mechanism in the SARS-CoV genome, and in particular to address the issues of the mutation rate and the time for the emergence of the SARS-CoV in the human population. We reported the estimated mutation rate in the SARS-CoV using the available complete genomic sequences whose clinical history either is certain or could be inferred. The sources of the genomic sequences used in this study and the methods of estimating mutation rates are presented in the Methods section. The divergence time was inferred based on the information summarized in Figure  1 . Table 1 shows the mutation rates estimated by three strategies. When the first strategy was used to adjust for sequencing errors and potential mutations in the cell culture, the mutation rate was estimated to be 0.80 -2.38 × 10 -3 nucleotide substitution per site per year using all the sequences not generated from mainland China, and 0.81 -1.38 × 10 -3 nucleotide substitution per site per year using the TOR2 and Urbani sequences only. When the second strategy was used, the mutation rate was estimated to be 0.74 -1.62 × 10 -3 nucleotide substitution per site per year, which is lower than that from using the first strategy. As expected, the mutation rate estimated using the third strategy was the lowest; 0.54 -1.57 × 10 -3 nucleotide substitution per site per year using the 11 sequences not generated from mainland China and 0.42 -0.72 × 10 -3 nucleotide substitution per site per year using the TOR2 and Urbani sequences only. For all samples, the proportion of non-synonymous substitutions per non-synonymous site (Ka) was 0.63 × 10 -3 and the proportion of synonymous substitutions per synonymous site (Ks) was 0.65 × 10 -3 , leading to Ka/Ks being 0.97. This ratio was 0.79 in the nonstructural region and 1.37 in the structural region. In particular, the values of Ka/Ks were 1.98 for nsp1 and 0.85 for S. Table 2 shows the rates of nucleotide substitution in the coding regions of sequences. The overall rates of non-synonymous and synonymous substitutions were 1.16 -3.30 × 10 -3 and 1.67 -4.67 × 10 -3 per site per year, respectively. The non-synonymous rate was higher in the three genes E, M, and N, suggesting some of those mutations might increase antigenicity, although the number of mutations used to calculate these rates was small. The mutation rate estimated earlier allowed us to estimate the age of the most recent common ancestor (MRCA) of the sample, which should be about the same or more recent than the time for the origin of SARS-CoV. The phylogeny reconstructed by the neighbor-joining method with mid-point rooting or by maximum parsimony is overall consistent with the epidemic (Additional file 1). All the sequences from mainland China clustered together and separated from the remaining sequences, including those clinically related to the index patient A. GZ01 was distantly separated from other sequences. Assuming the MRCA is the root of the phylogeny, the age of the MRCA Clinical relations and estimated range of the divergence time among 16 SARS-CoV isolates is then the divergence time between GZ01 and other sequences. Using the mutation rates estimated above, it is found that the MRCA could be alive at a time between March 28 and November 29, 2002 (strategy 1), between February 22 and October 3, 2002 (strategy 2), and even earlier (strategy 3). The most critical implication of these analyses is that it is entirely plausible that the MRCA of the sample could be alive as early as the spring of 2002. Some uncertainties in the quality of the sequence data and incomplete information from patient histories are two limiting factors of this study. The world-wide race to understand this novel virus has provided an unprecedented set of complete genome sequences of a pathogen in an interval of a few weeks, but likely side-effects of this race might be an elevated error rate in the released sequences and generating errors during the analysis. Among the 129 sequence variations reported [4] , many were generated randomly by the algorithms during the alignment of the multiple sequences, therefore these should be removed or adjusted. The concern above has led us to wait until all the sequences used in this study have been significantly revised by their generators and to manually adjust the multiple-sequence alignment. Still some errors were unavoidable partly due to the intrinsic error rate of sequencing technology. For example among 18 common variations, 9 could not be uniquely assigned to the internal branches of the phylogeny. This incongruence is likely partially due to sequence errors. The existence of sequence errors can also be inferred by examining the ratio of transitional versus transversional changes. If nucleotide substitution occurs randomly, there are two transversional substitutions on average for each transitional substitution, and the ratio of transition to transversion should be 0.5. However, transition is generally favored over transversion in many organisms. For example, the ratio is approximately 2 in the human genome [21, 22] . The ratio has not been discussed extensively in the RNA viruses; however, it appears to be higher than that in the mammalian genomes based on the two previous reports of 3.7 in the influenza A virus [23] and 5.0 in the Marburg virus [24] . In this study, 60 transitional substitutions and 54 transversional substitutions were observed among the 16 sequences, thus the ratio was 1.1. The ratio in five sequences from mainland China was 0.9, considerably smaller than 2.2 which was observed in the other eleven sequences. This suggests that sequences from mainland China may be more erroneous than the other sequences. On the other hand, the ratio was 0.9 for the In method 1, the nucleotide difference (3.2) among five Singaporean sequences was used to adjust the sequence errors and mutations that occurred during cell culture. In method 2, the number of variants between two sequences was reduced by 2 and the divergence time was increased by 14 days. In method 3, the nucleotide variants that were observed only once among the isolates were excluded. t = range of divergence time (days). µ = mutation rate (per site per year). singleton variations, which was much lower than the ratio of 3.5 for the non-singleton variants. This further indicates that singletons were more problematic. Because of the unknown level of errors in the sequences, a conservative approach to estimating the mutation rate was taken. Three strategies were used to reduce the effect of sequence errors, one being more aggressive than the other two. The mutation rates estimated by the first two strategies were quite similar. In the third strategy, all the variants unique to a given isolate were excluded. Such a strategy is very conservative because the amount of singletons is expected to be large in a rapid expanding environment (see below). Therefore the mutation rate was placed in the range of 0.80 -2.38 × 10 -3 nucleotide substitution per site per year based on the 11 sequences used. This rate, along with the rate of synonymous substitutions estimated in this study, is close to that recently reported using another approach [25] . In comparison to other coronaviruses, this rate is lower than that in the mouse hepatitis virus, similar to that in the transmissible gastroenteritis virus, but higher than that in the infectious bronchitis virus (Table 3 ) [6] [7] [8] . The estimated mutation rate is at the same order of magnitude as in other RNA viruses, for example, 2.3 × 10 -3 nucleotide substitution per site per year in the influenza A viruses [12, 13] . The estimated mutation rate in HIV appears to have a wide range [16, 17] . It is likely that the mutation rate in the SARS-CoV is not higher than that in HIV. Therefore, the SARS-CoV is not an unusual coronavirus or RNA virus in terms of its speed of nucleotide changes. One of the challenging tasks, therefore, is to find those variations which led to the SARS-CoV being unique from other RNA viruses, especially coronaviruses, and how those variations changed the functionality and helped to transmit it to humans. Nucleotide variation is distributed along the entire genome. Based on our alignment and the annotation in GenBank, 21 of the 26 open reading frames had the variations, including genes encoding polymerase, spike glycoprotein, envelope, membrane, and nucleocapsid protein. The estimated mutation rate suggests that approximately 2 to 6 new mutations will occur each month in a virus assuming the overall uniform mutation rate. However, the rate of the non-synonymous substitutions might vary during the course of the SARS-CoV evolution [25] . It was observed that there was an excess of mutations (and amino acid changes) in the external branches of the phylogeny of a large sample of the HA gene sequences of influenza A, which was partially caused by sampling bias [26] . From a population genetics standpoint, a large proportion of mutations should occur in the external branches when the infected hosts have rapidly increased. Therefore, one should not conclude that mutation rate is low because of a relatively small number of mutations in the internal branches [27] . Our analysis, even by a conservative estimation of mutation rate, indicates that the SARS-CoV population has already harbored a considerable amount of genetic diversity. [20] The emerging time of the SARS-CoV is of special importance in dissecting the origin of the virus as well as the dynamics of the epidemic. The time for the most recent common ancestor of the 16 isolates was estimated to be between February 2002 and November 2002. Although this is consistent with the date for the earliest known case of SARS and those estimated in other studies [25, 28] , it also suggests that SARS-CoV could have been present longer than generally believed, that is, around November 2002. One possible scenario is that the SARS-CoV had already infected some people in the spring of 2002 but failed to cause epidemics; its spread was however suppressed in the summer (similar to the summer of 2003), and re-emerged around November to cause the epidemic in 2003. Given the current re-emergence of SARS cases, this scenario is becoming more likely. There were indeed some media reports of SARS-like symptoms of patients in the spring of 2002 although none have been convincingly confirmed. An alternative scenario is that the common ancestor of the SARS-CoV lived in the spring of 2002, but the host was animals. The recent finding of high sequence homology between the isolate from a newly emerged SARS case (December 16, 2003) and the isolates from the masked palm civets [29] makes civets as the primary suspect of reservoir for SARS-CoV. The estimated mutation rate and the synonymous and non-synonymous substitution rates in the SARS-CoV genome were moderate compared to that in coronavirus and other RNA viruses, suggesting that the SARS-CoV is not an unusual coronavirus in terms of its speed of nucleotide or amino acid changes. Based on the mutation rates  We obtained 16 complete genomic sequences from the NCBI website http://www.ncbi.nlm.nih.gov/. Among them, five sequences (BJ01-04 and GZ01) were obtained from the hosts collected in mainland China and the remaining sequences (TOR2, Urbani, CUHK-W1, CUHK-Su10, HKU-39849, five Singaporean sequences, and TW1) were from the hosts in other geographic regions. Detailed information of the sequences is shown in Table  4 . CLUSTAL X [30] , a window-based user interface to the CLUSTAL W, was used to align the multiple sequences. The alignment was further manually examined and adjusted. All gene annotation information and nucleotide position designations in this study refer to the TOR2 sequence (GenBank accession ID: NC_004718). To avoid complications, only the single nucleotide variations were analyzed and all alignment gaps were excluded. This led to the identification of a total of 114 single nucleotide variations among all the sequences and an average of 18.2 nucleotide differences between two sequences. The MEGA2 computer program [31] was used to calculate the pair-wise nucleotide differences. The resulting genetic distances were corrected by Jukes and Cantor\\'s method [32] . The phylogeny of the sample was reconstructed using both neighbor-joining and maximum parsimony methods [31, 33] . Mutation rate can be estimated in principle by the number of nucleotide differences between two sequences divided by twice their divergent time, i.e., the time to their most recent common ancestor. Due to better documented contact histories, mutation rates were estimated only by the sequences whose hosts were not from mainland China, that is, sequences TOR2, Urbani, CUHK-W1, CUHK-Su10, HKU-39849, five Singaporean sequences, and TW1. First, the range of the divergence time between each pair of sequences was inferred based on information on infection history, reported strain isolation dates and sequence release dates (Additional file 2) [4, [34] [35] [36] . For example, the divergence time between isolates TOR2 and Urbani was estimated to be in the range of 34 to 58 days [35, 36] . Second, nucleotide difference between each pair of sequences was calculated with adjustments to reduce the effect of sequencing errors and potential mutations during cell culture. Three strategies were used. The first strategy was used to reduce the number of pair-wise nucleotide differences by the averaged number of nucleotide differences observed in five closely related Singaporean sequences [4] . This strategy effectively assumes that there is no real nucleotide difference among these five sequences so that their observed differences reflect the level of errors. The second strategy was used to reduce the pair-wise nucleotide difference by two and to add 7 days to the divergence time to account for cell culture time. This strategy assumes that the mutation rate during the cell culture is the same as that in the human host and that on average the sequencing error is one nucleotide per genome. In the third strategy, we excluded all the nucleotide variants which had been observed only once (singletons) among the 61 human SARS-CoV sequences reported in [25] . The rational is that non-singleton mutations observed in a sample are much less likely due to sequencing errors as well as mutations during the laboratory passage of virus. This strategy is apparently conservative and can be regarded as the lower bound of the mutation rate. Finally, the mutation rate per site per year was estimated by where d ij is the genetic distance between sequence i and j, t ij is twice their divergence time (in number of days), and n is the number of sequences. A mutation in a codon is non-synonymous (or nonsilent) if it changes the amino acid, and is synonymous (silent) otherwise. The number of non-synonymous mutations per non-synonymous site (Ka) and the number of synonymous mutations per synonymous site (Ks) were computed using the method of Li, Wu, and Luo [37] . The non-synonymous and synonymous substitution rates were calculated using the divergence time as estimated above. Only the second strategy was applied to the rate estimation because the number of nucleotide differences used for the adjustment in the first strategy can not be separated for the non-synonymous and synonymous mutations. '"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[16188][\"text\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(47298, 19)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
